File: recursion/javascript/solutions/smallest.js
/**
 * Solution for smallest.
 */

const { Node, isEmpty, unprepend } = require('./linkedList');

function smallest(list) {
  if (isEmpty(list.next)) {
    return list.value;
  }

  let [first, rest] = unprepend(list);

  return Math.min(first, smallest(rest));
}

if (require.main === module) {
  console.assert(smallest(Node.fromArray([3, 7, 2])) === 2);
  console.assert(smallest(Node.fromArray([1, 2, 3, 4, 5])) === 1);
  console.assert(smallest(Node.fromArray([5, 4, 3, 2, 1])) === 1);
  console.assert(smallest(Node.fromArray([42])) === 42);
  console.assert(smallest(Node.fromArray([-5, -3, -10])) === -10);

  console.log('All tests passed!');
}

module.exports = { smallest };

File: recursion/python/solutions/smallest.py
"""Solution for smallest."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend


def smallest(lst):
    """
    Given a non-empty list of numbers, return the smallest.

    smallest(3 -> 7 -> 2) => 2
    """
    if is_empty(lst.next):
        return lst.value

    first, rest = unprepend(lst)

    return min(first, smallest(rest))


if __name__ == '__main__':
    # Test cases
    assert smallest(Node.from_list([3, 7, 2])) == 2
    assert smallest(Node.from_list([1, 2, 3, 4, 5])) == 1
    assert smallest(Node.from_list([5, 4, 3, 2, 1])) == 1
    assert smallest(Node.from_list([42])) == 42
    assert smallest(Node.from_list([-1, -5, -2])) == -5
    assert smallest(Node.from_list([0, 0, 0])) == 0

    print("All tests passed!")
